File: clients/node/src/vlp.ts
// VLP v0.2 wire encode / decode + CRC-32C.
//
// The normative wire specification lives at `book/src/spec/vlp.md`. This
// module is the canonical Node.js implementation that the production
// `Varta` client builds on top of. Cross-language byte equality is
// enforced by the conformance test suite at `tests/conformance.test.ts`,
// which drives `tools/vlp-test-vectors.json`.

export const MAGIC: Buffer = Buffer.from([0x56, 0x41]); // "VA"
export const VERSION: number = 0x02;
export const NONCE_TERMINAL: bigint = 0xffffffffffffffffn;
export const TIMESTAMP_INVALID: bigint = 0xffffffffffffffffn;
export const FRAME_BYTES: number = 32;

// Beat status — matches the 1-byte wire value at offset 3.
// `Stall` is observer-synthesized; agents MUST NOT emit it on the wire.
// `decode()` raises `DecodeError` with `kind="StallOnWire"` if it appears.
export enum Status {
  Ok = 0,
  Degraded = 1,
  Critical = 2,
  Stall = 3,
}

export type StatusLike = Status | number | string;

const STATUS_BY_NAME: Record<string, Status> = {
  ok: Status.Ok,
  degraded: Status.Degraded,
  critical: Status.Critical,
  stall: Status.Stall,
};

// Decode error kinds — these strings MUST match the entries in
// `tools/vlp-test-vectors.json` and the corresponding constants in
// the Rust, Python, and Go clients (cerebrum 2026-05-13).
export type DecodeErrorKind =
  | "BadMagic"
  | "BadVersion"
  | "BadCrc"
  | "BadStatus"
  | "StallOnWire"
  | "BadPid"
  | "BadTimestamp"
  | "BadNonce";

export class DecodeError extends Error {
  readonly kind: DecodeErrorKind;
  constructor(kind: DecodeErrorKind, detail = "") {
    super(detail ? `${kind}: ${detail}` : kind);
    this.name = "DecodeError";
    this.kind = kind;
  }
}

// CRC-32C (Castagnoli) — RFC 3720 appendix B. Reflected polynomial
// 0x82F63B78, init 0xFFFFFFFF, refin/refout, xorout 0xFFFFFFFF. Matches
// the canonical Rust implementation at `crates/varta-vlp/src/crc32c.rs`.

const POLY_REFLECTED = 0x82f63b78;

const CRC_TABLE: Uint32Array = (() => {
  const table = new Uint32Array(256);
  for (let i = 0; i < 256; i++) {
    let c = i >>> 0;
    for (let k = 0; k < 8; k++) {
      c = (c & 1) !== 0 ? (c >>> 1) ^ POLY_REFLECTED : c >>> 1;
    }
    table[i] = c >>> 0;
  }
  return table;
})();

export function crc32c(data: Buffer | Uint8Array): number {
  let crc = 0xffffffff >>> 0;
  for (let i = 0; i < data.length; i++) {
    const idx = ((crc ^ data[i]!) & 0xff) >>> 0;
    crc = ((CRC_TABLE[idx]! ^ (crc >>> 8)) >>> 0);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

// Decoded view of a 32-byte VLP v0.2 frame. Mirrors the Python
// `Frame` dataclass and Go `Frame` struct field-for-field.
export interface Frame {
  status: Status;
  pid: number;
  timestamp: bigint;
  nonce: bigint;
  payload: number;
}

function coerceStatus(status: StatusLike): Status {
  if (typeof status === "string") {
    const s = STATUS_BY_NAME[status.toLowerCase()];
    if (s === undefined) {
      throw new RangeError(`unknown status name: ${status}`);
    }
    return s;
  }
  if (status === Status.Ok || status === Status.Degraded ||
      status === Status.Critical || status === Status.Stall) {
    return status as Status;
  }
  throw new RangeError(`invalid status: ${status}`);
}

// Encode a single VLP v0.2 frame into a fresh 32-byte Buffer.
//
// The hot path of the production client uses `encodeInto` against a
// pre-allocated scratch buffer to keep per-beat allocation minimal.
export function encode(
  status: StatusLike,
  pid: number,
  timestamp: bigint,
  nonce: bigint,
  payload: number,
): Buffer {
  const buf = Buffer.alloc(FRAME_BYTES);
  encodeInto(buf, status, pid, timestamp, nonce, payload);
  return buf;
}

// Encode a single VLP v0.2 frame into `buf` (must be ≥ 32 bytes).
export function encodeInto(
  buf: Buffer,
  status: StatusLike,
  pid: number,
  timestamp: bigint,
  nonce: bigint,
  payload: number,
): void {
  if (buf.length < FRAME_BYTES) {
    throw new RangeError(`buffer must be at least ${FRAME_BYTES} bytes`);
  }
  const s = coerceStatus(status);
  buf[0] = 0x56;
  buf[1] = 0x41;
  buf[2] = VERSION;
  buf[3] = s & 0xff;
  buf.writeUInt32LE(pid >>> 0, 4);
  buf.writeBigUInt64LE(timestamp, 8);
  buf.writeBigUInt64LE(nonce, 16);
  buf.writeUInt32LE(payload >>> 0, 24);
  const crc = crc32c(buf.subarray(0, 28));
  buf.writeUInt32LE(crc, 28);
}

// Decode a 32-byte VLP v0.2 frame.
//
// Throws `DecodeError` on the first failed validation step. See
// `book/src/spec/vlp.md` §5 for the normative decode order:
// magic → version → CRC → status (incl. Stall rejection) → pid →
// timestamp → nonce.
export function decode(buf: Buffer | Uint8Array): Frame {
  if (buf.length !== FRAME_BYTES) {
    throw new DecodeError("BadMagic", `length ${buf.length} != ${FRAME_BYTES}`);
  }
  if (buf[0] !== 0x56 || buf[1] !== 0x41) {
    throw new DecodeError(
      "BadMagic",
      Buffer.from(buf.slice(0, 2)).toString("hex"),
    );
  }
  if (buf[2] !== VERSION) {
    throw new DecodeError(
      "BadVersion",
      `0x${(buf[2]!).toString(16).padStart(2, "0")}`,
    );
  }

  const view = Buffer.isBuffer(buf) ? buf : Buffer.from(buf);
  const storedCrc = view.readUInt32LE(28);
  const computedCrc = crc32c(view.subarray(0, 28));
  if (storedCrc !== computedCrc) {
    throw new DecodeError(
      "BadCrc",
      `expected ${computedCrc.toString(16).padStart(8, "0")}, got ${storedCrc
        .toString(16)
        .padStart(8, "0")}`,
    );
  }

  const statusByte = view[3]!;
  if (statusByte > 3) {
    throw new DecodeError(
      "BadStatus",
      `0x${statusByte.toString(16).padStart(2, "0")}`,
    );
  }
  if (statusByte === Status.Stall) {
    throw new DecodeError("StallOnWire");
  }

  const pid = view.readUInt32LE(4);
  const timestamp = view.readBigUInt64LE(8);
  const nonce = view.readBigUInt64LE(16);
  const payload = view.readUInt32LE(24);

  if (pid === 0 || pid === 1) {
    throw new DecodeError("BadPid", String(pid));
  }
  if (timestamp === TIMESTAMP_INVALID) {
    throw new DecodeError("BadTimestamp");
  }
  if (nonce === NONCE_TERMINAL && statusByte !== Status.Critical) {
    throw new DecodeError(
      "BadNonce",
      `nonce=NONCE_TERMINAL paired with status=0x${statusByte
        .toString(16)
        .padStart(2, "0")}`,
    );
  }

  return {
    status: statusByte as Status,
    pid,
    timestamp,
    nonce,
    payload,
  };
}

// Convenience wrapper for parity with the Go client's exported
// `DecodeFrame`. Returns the same shape as `decode()`.
export function decodeFrame(buf: Buffer | Uint8Array): Frame {
  return decode(buf);
}

File: clients/python/src/varta/_vlp.py
"""VLP v0.2 wire encode / decode + CRC-32C.

The normative wire specification lives at ``book/src/spec/vlp.md``. This
module is the canonical Python implementation that the production
:class:`varta.Varta` client builds on top of. The verifier-grade twin
at ``tools/reference-implementations/python/vlp.py`` is kept in step but
exists for a different stability contract (pinned to the spec version,
not to this package's semver).

Standard library only. Python 3.8+.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Tuple, Union

__all__ = [
    "MAGIC",
    "VERSION",
    "NONCE_TERMINAL",
    "FRAME_BYTES",
    "Status",
    "DecodeError",
    "Frame",
    "crc32c",
    "encode",
    "encode_into",
    "decode",
]

MAGIC: bytes = b"\x56\x41"  # "VA"
VERSION: int = 0x02
NONCE_TERMINAL: int = 0xFFFFFFFFFFFFFFFF
FRAME_BYTES: int = 32


class Status(IntEnum):
    """Beat status — matches the 1-byte wire value at offset 3.

    ``STALL`` is observer-synthesized; agents MUST NOT emit it on the wire.
    :class:`DecodeError` with ``kind="StallOnWire"`` is raised if it appears.
    """

    OK = 0
    DEGRADED = 1
    CRITICAL = 2
    STALL = 3


_STATUS_BY_NAME = {
    "ok": Status.OK,
    "degraded": Status.DEGRADED,
    "critical": Status.CRITICAL,
    "stall": Status.STALL,
}


StatusLike = Union[Status, int, str]


class DecodeError(Exception):
    """Raised on any wire-format validation failure.

    The ``kind`` attribute is the spec-defined error variant name, which
    matches the strings in ``tools/vlp-test-vectors.json``:

        BadMagic, BadVersion, BadCrc, BadStatus, StallOnWire,
        BadPid, BadTimestamp, BadNonce.
    """

    __slots__ = ("kind",)

    def __init__(self, kind: str, detail: str = "") -> None:
        super().__init__(f"{kind}: {detail}" if detail else kind)
        self.kind = kind


# ---------------------------------------------------------------------------
# CRC-32C (Castagnoli) — RFC 3720 appendix B.
# ---------------------------------------------------------------------------

_POLY_REFLECTED = 0x82F63B78


def _build_crc_table() -> Tuple[int, ...]:
    table = []
    for i in range(256):
        c = i
        for _ in range(8):
            c = (c >> 1) ^ _POLY_REFLECTED if c & 1 else c >> 1
        table.append(c)
    return tuple(table)


_CRC_TABLE: Tuple[int, ...] = _build_crc_table()


def crc32c(data: bytes) -> int:
    """Compute the CRC-32C (Castagnoli) checksum of ``data``.

    Reflected polynomial 0x82F63B78, init 0xFFFFFFFF, refin/refout,
    xorout 0xFFFFFFFF. Matches RFC 3720 appendix B and the canonical
    Rust implementation at ``crates/varta-vlp/src/crc32c.rs``.
    """
    crc = 0xFFFFFFFF
    for b in data:
        crc = _CRC_TABLE[(crc ^ b) & 0xFF] ^ (crc >> 8)
    return crc ^ 0xFFFFFFFF


# ---------------------------------------------------------------------------
# Frame
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Frame:
    """Decoded view of a 32-byte VLP v0.2 frame."""

    status: Status
    pid: int
    timestamp: int
    nonce: int
    payload: int


def _coerce_status(status: StatusLike) -> Status:
    if isinstance(status, str):
        try:
            return _STATUS_BY_NAME[status.lower()]
        except KeyError as exc:
            raise ValueError(f"unknown status name: {status!r}") from exc
    return Status(status)


def encode(
    status: StatusLike,
    pid: int,
    timestamp: int,
    nonce: int,
    payload: int,
) -> bytes:
    """Encode a single VLP v0.2 frame into a fresh ``bytes`` of length 32.

    The hot path of the production client uses :func:`encode_into` against
    a pre-allocated scratch buffer to keep per-beat allocation minimal.
    """
    buf = bytearray(FRAME_BYTES)
    encode_into(buf, status, pid, timestamp, nonce, payload)
    return bytes(buf)


def encode_into(
    buf: bytearray,
    status: StatusLike,
    pid: int,
    timestamp: int,
    nonce: int,
    payload: int,
) -> None:
    """Encode a single VLP v0.2 frame into ``buf`` (must be ``len == 32``).

    Reuses the caller's buffer so steady-state beats avoid allocating a
    fresh frame object per emission. Python cannot match the Rust client's
    zero-heap guarantee — :mod:`struct` returns new ``bytes`` internally —
    but reusing the destination buffer at least eliminates the wrapper
    allocation.
    """
    if len(buf) != FRAME_BYTES:
        raise ValueError(f"buffer must be exactly {FRAME_BYTES} bytes")
    status_int = int(_coerce_status(status))
    struct.pack_into(
        "<2sBBIQQI",
        buf,
        0,
        MAGIC,
        VERSION,
        status_int,
        pid,
        timestamp,
        nonce,
        payload,
    )
    struct.pack_into("<I", buf, 28, crc32c(bytes(buf[:28])))


def decode(buf: bytes) -> Frame:
    """Decode a 32-byte VLP v0.2 frame.

    Raises :class:`DecodeError` on the first failed validation step. See
    ``book/src/spec/vlp.md`` §5 for the normative decode order:
    magic → version → CRC → status (incl. Stall rejection) → pid →
    timestamp → nonce.
    """
    if len(buf) != FRAME_BYTES:
        raise DecodeError("BadMagic", f"length {len(buf)} != {FRAME_BYTES}")
    if buf[0:2] != MAGIC:
        raise DecodeError("BadMagic", buf[0:2].hex())
    if buf[2] != VERSION:
        raise DecodeError("BadVersion", f"0x{buf[2]:02x}")

    (stored_crc,) = struct.unpack("<I", buf[28:32])
    computed_crc = crc32c(buf[0:28])
    if stored_crc != computed_crc:
        raise DecodeError(
            "BadCrc", f"expected {computed_crc:08x}, got {stored_crc:08x}"
        )

    status_byte = buf[3]
    if status_byte not in (Status.OK, Status.DEGRADED, Status.CRITICAL, Status.STALL):
        raise DecodeError("BadStatus", f"0x{status_byte:02x}")
    if status_byte == Status.STALL:
        raise DecodeError("StallOnWire")

    (pid,) = struct.unpack("<I", buf[4:8])
    (timestamp,) = struct.unpack("<Q", buf[8:16])
    (nonce,) = struct.unpack("<Q", buf[16:24])
    (payload,) = struct.unpack("<I", buf[24:28])

    if pid in (0, 1):
        raise DecodeError("BadPid", str(pid))
    if timestamp == 0xFFFFFFFFFFFFFFFF:
        raise DecodeError("BadTimestamp")
    if nonce == NONCE_TERMINAL and status_byte != Status.CRITICAL:
        raise DecodeError(
            "BadNonce",
            f"nonce=NONCE_TERMINAL paired with status=0x{status_byte:02x}",
        )

    return Frame(
        status=Status(status_byte),
        pid=pid,
        timestamp=timestamp,
        nonce=nonce,
        payload=payload,
    )
